### PR TITLE
Support for Dynamic Pricing in Daemon 

### PR DESCRIPTION
**Pull request by anandrgitnirman**
PS see #526 for details

### CODE REVIEW

File: handler/grpc.go [hunk 1]
```diff
@@ -252,6 +266,59 @@ func (g grpcHandler) grpcToJSONRPC(srv interface{}, inStream grpc.ServerStream)
...
+	sentMessage      interface{}
+}
+
+func (f *wrapperServerStream) SetTrailer(md metadata.MD) {
+	f.stream.SetTrailer(md)
+}
+
+func NewWrapperServerStream(stream grpc.ServerStream) (grpc.ServerStream, error) {
+	m := &codec.GrpcFrame{}
+	err := stream.RecvMsg(m)
```
Comment (vsbogd) [inline handler/grpc.go:282]: Why receiving message is needed here?

File: handler/interceptors.go [hunk 2]
```diff
@@ -235,10 +244,72 @@ func GrpcPaymentValidationInterceptor(defaultPaymentHandler PaymentHandler, paym
 }
 
 type paymentValidationInterceptor struct {
+	serviceMetadata       *blockchain.ServiceMetadata
 	defaultPaymentHandler PaymentHandler
 	paymentHandlers       map[string]PaymentHandler
 }
 
+func (interceptor *paymentValidationInterceptor) checkForDynamicPricing(srv interface{}, ss grpc.ServerStream,
```
Comment (vsbogd) [inline handler/interceptors.go:252]: My original intention was isolating pricing related logic in `pricing` module. In particular such implementation will allow get price dynamically only when it is needed (i.e. don't do this when `free-call` or `authorized-user` payment strategy is applied).

If I would implement this I would do the following:
- move checking if dynamic pricing is applicable to `PricingStrategy.determincePricingApplicable` method
- add new `DYNAMIC_PRICING` into `PriceType`
- add new `dynamic_method_pricing.go` and move getting price related code into it (similar to `fixed_method_pricing.go`
- to make above possible probably we need to modify `handler.GrpcStreamContext` and add additional information into it if needed
- `ServiceMetadata` should be part of the `PricingStrategy` I believe, but I am not sure how to implement it better

File: handler/grpc.go (hunk 1 above)
Comment (anandrgitnirman) [reply handler/grpc.go:282]: the same message will be used to make a call to estimate price and also to the actual service invocation

File: handler/interceptors.go (hunk 2 above)
Comment (anandrgitnirman) [reply handler/interceptors.go]: made changes , if it looks fine , will check in the test cases as well

File: handler/grpc.go [hunk 3]
```diff
@@ -174,7 +175,20 @@ func forwardServerToClient(src grpc.ServerStream, dst grpc.ClientStream) chan er
...
+			//future calls will result in io.EOF , we want to retrieve the
+			//the first message sent by the client and pass this on the regualar service call
+			//This is done to be able to make calls to support regualr Service call + Dynamic pricing call
+			if i == 0 {
+				//todo we need to think through to determine price for every call on stream calls
+				//will be handled when we support streaming and pricing across all clients in snet-platform
+				if wrappedStream, ok := src.(*wrapperServerStream); ok {
+					f = (wrappedStream.OriginalRecvMsg()).(*codec.GrpcFrame)
+				} else {
+					src.RecvMsg(f)
```
[the remark this answers is not in the document]
Comment (anandrgitnirman) [reply handler/grpc.go]: fixed this

**Review by vsbogd — APPROVED** [review]

**Comment by vsbogd** [comment]
Looks good to me, thanks

**Pull request merged**